### ISSUE TITLE
Remote emojis

### ISSUE THREAD

**Issue by roipoussiere** [tootsuite/mastodon#8649]
I love custom emojis, it puts some creativity and fun in the fediverse.

The drawback is that we depend in our instance admin to use them.

I regularly see some toots with amazing emojis that I really want to use, but I don't want to spam my admin for this. And everyone in my instance have their own feelings and interests, so even if my admin is very nice (of course he is), I don't want to have 100+ custom emojis on my emojis selector. Also, some custom emojis are good to use just one time for a specific context and it's not really necessary to store them all the time.

**We could fix this with remote emojis: the ability to insert an emoji stored in another instance.**

The syntax for custom emojis is `:emoji_name:`. It could be extended to `:emoji_name@mastodon_server:`.

Later, the front-end could be improved to put this syntax in our toot when we copy-paste a text containing a custom emoji from a remote server, to improve usability. But this UX-related suggestion could be the subject of an other issue when / if the backend of this one is merged.

Note that only local emojis will be shown in the custom emoji list in emoji selector: in this way, we still mainly use our local emojis.

At this point, it's not so difficult for some administrator to write a script who analyses local toots, looking for all remote emojis to get some statistics on them, in order to add the most used ones.

In the same way, admins could analyses emojis call in the instance web server logs and if some emoji are used very often, they ask for other admins (eventually, admins of instances where most of the logs comes from): "Hey, could you add this one please?".

---

In fact, I have a fun project in mind waiting for this feature, I allow myself to share it here to give you an example of the possibilities provided by remote emojis: an emoji bot. Someone send a toot to the bot, with an image or gif on it, and eventually several hashtags categorizing the image, and the bot adds the emoji on its dedicated instance on the fly, and answers to the toot with that emoji. The server will also provides a web page displaying all current stored emoji, and we can search them by entering keywords (according to the keywords used when sending the emoji to to bot).

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by bonsaiheldin** [comment]
Funny how emojis are so so so important today. 😅

**Comment by 00dani** [comment]
Chiming in to say that I independently proposed the `:emoji_name@instance.domain:` syntax [a few minutes ago](https://vulpine.club/@00dani/100710851960922667), so I think we can conclude it's a fairly natural syntax for doing this! :wink:

**Comment by alex-coding-c** [comment]
:yell@mastodon.social:

**Comment by alex-coding-c** [comment]
:yell: As you can see the yell emoji is not available on this github but in mastodon.  I'd like to show it on the pleroma but not sure how to resolve it.

**Comment by ArisuOngaku** [comment]
This is an important feature imho.

**Comment by phseiff** [comment]
It'd also really be in the spirit of the fediverse, I think, to have support for "federated emojis" rather than having to choose between local unfederated emojis and centralized (by the Emoji standard, I mean) emojis.

**Comment by HappiePlant** [comment]
It's like Discord Nitro for free

**Comment by EasternPA** [comment]
So what's the latest on this? I can see the remote emojis on my Admin panel, which are all enabled. Is there a way to reference them for local use, or do I need to download each one and re-upload it to my instance as local only? I'm sure I'm missing something simple. It must be that.

**Comment by willfanguy** [comment]
For those still trying to figure this out, I've outlined in the attached image how to copy custom emoji (that show up as remote like @EasternPA mentioned above) over to your local instance. Hope this helps!

![CleanShot 2022-12-01 at 13 47 58@2x](https://user-images.githubusercontent.com/12038046/205146148-a710daf0-e12a-4723-84f4-a011182b9c84.png)

**Comment by Andre601** [comment]
I mentioned this in the above linked issue, but repeat it here too:

Some instance owners may have an interest in not having their emojis used on other instances... There could be multiple reasons (Bandwidth usage, privacy(?), ...), but an option to toggle if custom emojis should be usable by other instances should be considered here.

**Comment by foxyseta** [comment]
Even though the `:emoji_name@mastodon_server:` looks very natural and also similar to user handles, won't it make autocomplete impossible? When typing `emoj`, how does the client know which instance(s) to use for the search? Something like `:mastodon_server/emoji_name:` would allow the client to offer autocomplete at `:mastodon_s` from the list of known servers (I think?), and at `:mastodon_server/emoj` from th emojis endpoint of the selected server.

**Comment by KaKi87** [comment]
Indeed, although `:emoji:` should remain usable for local ones, so to input a server, `@` should still be used, e.g. `:@server/emoji:`.

**Comment by Andre601** [comment]
I feel like auto-complete should remain limited to the instance you use here... After all, with countless instances out there, it could become quite overwhelming to find the one you would like to use...

**Comment by foxyseta** [comment]
> I feel like auto-complete should remain limited to the instance you use here... After all, with countless instances out there, it could become quite overwhelming to find the one you would like to use...

Well, the client I am using does offer it, so that's why I was considering those kind of clients, too. Of course, this is a client-specific feature, so you are always free to use a client that does not implement it or allows you to turn it off. I also think think is limited to servers known to your own instance, anyway.
